### PR TITLE
NIFI-12441 Added No Tracking listing strategy to ListS3

### PR DESCRIPTION
**Pull request by juldrixx**
# Summary

[NIFI-12441](https://issues.apache.org/jira/browse/NIFI-12441)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [X] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [X] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [X] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [X] Pull Request based on current revision of the `main` branch
- [X] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [X] Build completed using `mvn clean install -P contrib-check`
  - [X] JDK 21

### Licensing

- [X] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [X] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [X] Documentation formatting appears as expected in rendered files

### CODE REVIEW

File: nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/ListS3.java [hunk 1]
```diff
@@ -451,11 +456,92 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
...
+        final Long maxAgeMilliseconds = context.getProperty(MAX_AGE) != null ? context.getProperty(MAX_AGE).asTimePeriod(TimeUnit.MILLISECONDS) : null;
+        final long listingTimestamp = System.currentTimeMillis();
+
+        final String bucket = context.getProperty(BUCKET_WITHOUT_DEFAULT_VALUE).evaluateAttributeExpressions().getValue();
+        final int batchSize = context.getProperty(BATCH_SIZE).asInteger();
+
+        int listCount = 0;
+        int totalListCount = 0;
+
+        getLogger().trace("Start listing, listingTimestamp={}", new Object[]{listingTimestamp});
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/ListS3.java:482]: The `Object[]` wrapper is not necessary for log arguments, although not all components have been updated to reflect the recommended usage, new code should follow the pattern.
```suggestion
        getLogger().trace("Start listing, listingTimestamp={}", listingTimestamp);
```

File: nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/ListS3.java [hunk 2]
```diff
@@ -451,11 +456,92 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
...
+            do {
+                VersionListing versionListing = bucketLister.listVersions();
+                for (S3VersionSummary versionSummary : versionListing.getVersionSummaries()) {
+                    long lastModified = versionSummary.getLastModified().getTime();
+                    if ((maxAgeMilliseconds != null && (lastModified < (listingTimestamp - maxAgeMilliseconds)))
+                        || lastModified > (listingTimestamp - minAgeMilliseconds)) {
+                        continue;
+                    }
+
+                    getLogger().trace("Listed key={}, lastModified={}", new Object[]{versionSummary.getKey(), lastModified});
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/ListS3.java:504]:
```suggestion
                    getLogger().trace("Listed key={}, lastModified={}", versionSummary.getKey(), lastModified);
```

File: nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/ListS3.java [hunk 3]
```diff
@@ -451,11 +456,92 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
...
+                    getLogger().info("Successfully listed {} new files from S3; routing to success", listCount);
+                    session.commitAsync();
+                }
+
+                listCount = 0;
+            } while (bucketLister.isTruncated());
+
+            writer.finishListing();
+        } catch (final Exception e) {
+            getLogger().error("Failed to list contents of bucket due to {}", e, e);
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/ListS3.java:529]: The `due to {}` approach is duplicative and also needs to be removed from older implementations, but this is an opportunity to streamline the log and avoid duplicating the message, as it is already included in the stack trace.
```suggestion
            getLogger().error("Failed to list contents of bucket", e);
```

File: nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/ListS3.java [hunk 4]
```diff
@@ -451,11 +456,92 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
...
+            writer.finishListingExceptionally(e);
+            session.rollback();
+            context.yield();
+            return;
+        }
+
+        final long listMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNanos);
+        getLogger().info("Successfully listed S3 bucket {} in {} millis", new Object[]{bucket, listMillis});
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/ListS3.java:537]:
```suggestion
        getLogger().info("Successfully listed S3 bucket {} in {} millis", bucket, listMillis);
```

File: nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/ListS3.java [hunk 5]
```diff
@@ -451,11 +456,92 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
...
+
+        if (totalListCount == 0) {
+            getLogger().debug("No new objects in S3 bucket {} to list. Yielding.", new Object[]{bucket});
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/ListS3.java:540]:
```suggestion
            getLogger().debug("No new objects in S3 bucket {} to list. Yielding", bucket);
```

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
Thanks for the improvement @juldrixx. The general approach looks good based on other implementations, but I noted a few syntax recommendations based on current approaches.

**Comment by exceptionfactory** [comment]
@juldrixx There appears to be a test failure in the `ci-workflow` for the new listing strategy:

```
org.apache.nifi.processors.aws.s3.TestListS3.testNoTrackingList -- Time elapsed: 0.020 s <<< FAILURE
```

**Comment by juldrixx** [comment]
> @juldrixx There appears to be a test failure in the `ci-workflow` for the new listing strategy:
> 
> ```
> org.apache.nifi.processors.aws.s3.TestListS3.testNoTrackingList -- Time elapsed: 0.020 s <<< FAILURE
> ```

Sorry, I didn't see that the unit test results had changed when I rebased. It should be fine now.

**Review by exceptionfactory — APPROVED** [review]
Thanks for making the adjustments and corrections @juldrixx, the latest version looks good! +1 merging

**Pull request closed without merging**